layout event refactor

tomnelson opened the pull request:
Changes to be committed:
	new file:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/event/LayoutChange.java
	new file:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/event/LayoutNodePositionChange.java
	new file:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/event/LayoutStateChange.java
	modified:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/AbstractLayoutModel.java
	deleted:    jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/DefaultLayoutModelChangeSupport.java
	deleted:    jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/DefaultLayoutStateChangeSupport.java
	modified:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/LayoutModel.java
	modified:   jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/LoadingCacheLayoutModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BaseVisualizationModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/AggregateLayoutModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/AbstractSpatial.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialQuadTree.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialRTree.java

tomnelson commented:
This PR wrestled all of the layout related event systems into similarly designed containers, with documentation telling what they do, and telling how they are used by consumers in jung-visualization.

tomnelson commented on jung-visualization/src/test/java/edu/uci/ics/jung/visualization/spatial/rtree/RTreeTest.java:
```diff
@@ -227,25 +229,23 @@ private void testAreas(TreeNode rootNode) {
+          .usingTolerance(1.0E-4)
+          .equals(unionDimensions);
+    }
     for (TreeNode rt : rootNode.getChildren()) {
```
I put the 4 doubles {x, y, w, h} into arrays and compared the arrays with a tolerance. It seems like that is more straightforward than building a custom Subject to compare java.awt.Shape instances. Thoughts?

jbduncan replied:
It's not entirely clear to me either if it's worth creating a custom `Subject` for `java.awt.Shape` at this stage, so I agree. :)

tomnelson commented:
It's not obvious from looking only at the diffs, but LayoutModel and AbstractLayoutModel are greatly cleaned up by the removal of all the event mode implementation details.

tomnelson commented on jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialQuadTree.java:
```diff
@@ -541,15 +539,13 @@ public void update(N node, Point location) {
-    this.update(evt.getNode(), evt.getLocation());
+  public void layoutNodePositionChanged(LayoutNodePositionChange.NetworkEvent<N> evt) {
+    this.update(evt.node, evt.location);
   }
```
this event allows the update of one element in the tree (vice rebuilding the whole thing)

tomnelson commented on jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialRTree.java:
```diff
@@ -298,18 +296,18 @@ public void recalculate() {
   }
 
   public static class Edges<E, N> extends SpatialRTree<E, N>
-      implements Spatial<E>, LayoutChangeListener<N> {
```
not sure if it was clear why this is <E, N>.
The element shapes stored in the tree are the edges, so the rtree is a collection of E shapes (as opposed to N shapes). In order to work with the edges, the tree also needs access to the Node endpoints. Thus, the SpatialRTree.Edges<E, N> implements Spatial<E>
(an RTree of E shapes with a layout model of N things)
and
SpatialRTree.Nodes<N, N> implements Spatial<N>
(an RTree of N shapes with a layoutMode of N things)

jbduncan commented on jung-algorithms/src/test/java/edu/uci/ics/jung/layout/spatial/FRLayoutsTest.java:
```diff
@@ -109,9 +111,11 @@ public static void check() {
+    assertWithMessage(
+            "The compared maps are not close enough: mapTwo: %s, mapThree: %s", mapTwo, mapThree)
+        .that(mapTwo)
+        .comparingValuesUsing(Correspondence.tolerance(1.0E-4))
```
Have you had a read through [this page](https://google.github.io/truth/floating_point#for-approximate-equality-what-tolerance-should-i-use) on the Truth website about the sort of general-rule-of-thumb tolerance you should be aiming to test against? I ask because although I think that a tolerance of `1.0E-4` looks pretty reasonable, I also think you might be able to reduce it to a smaller number if the values in `mapTwo` tend to be small.

tomnelson replied:
On the viz side, they all become pixel integers, so almost any tolerance will do. I'm really just looking for cases where they are way off.

jbduncan replied:
Okay, cool, no problems for me here then. A comment would be good, I think, for future maintainers, but I won't insist. :)

jbduncan commented on jung-algorithms/src/test/java/edu/uci/ics/jung/layout/spatial/FRLayoutsTest.java:
```diff
@@ -109,9 +111,11 @@ public static void check() {
+            "The compared maps are not close enough: mapTwo: %s, mapThree: %s", mapTwo, mapThree)
+        .that(mapTwo)
+        .comparingValuesUsing(Correspondence.tolerance(1.0E-4))
+        .equals(mapThree);
```
I think this invocation of `.equals` won't do as you expect, because `Subject.equals` does the same thing as `Object.equals` (i.e. return true if `mapTwo` is equal to `mapThree`; false otherwise), whereas what I think you're after is functionality like `Assert.assertEquals`. Is my observation correct?

If so, you should replace this and all other calls to `Subject.equals` with `isEqualTo`. The JavaDoc for `Subject.equals` describes that:

> `Object.equals(Object)` is not supported on Truth subjects. If you meant to test object equality between an expected and the actual value, use `isEqualTo(Object)` instead.

tomnelson replied:
Good catch.
I was lulled into thinking it was easier than it is. For one, I am comparing Point instances, so I would probably need a Subject for Point. The compiler will not (and should not) let me just swap out isEqualTo and equals.

jbduncan replied:
Maybe error-prone would catch this at compile-time, if/once we start using it? I don't know if error-prone catches anything Truth-related, but the error-prone team have implemented checks specifically for Google-originated open source projects, so I wouldn't be surprised if a check for this appears eventually if it hasn't yet.

jbduncan commented on jung-visualization/pom.xml:
```diff
@@ -14,6 +14,12 @@
+      <artifactId>truth</artifactId>
+      <version>0.39</version>
+      <scope>test</scope>
+    </dependency>
```
I think that the version information for Truth here (i.e. the fact that we're using version `0.39`), as well as the version info in all other poms where you introduced Truth, could be externalised to the root `pom.xml`. So I'd expect to see the following property:

```xml
<truth.version>0.39</truth.version>
```

added just right underneath [these lines](https://github.com/jrtom/jung/blob/master/pom.xml#L68-L70) in the root pom, and then replacing all usages of `0.39` in the child poms with `${truth.version}`. So where you would have had the following sort of line previously:

```xml
<version>0.39</version>
```

it would now look like:

```xml
<version>${truth.version}</version>
```

Have I described myself well?

tomnelson replied:
I figured that would come in via your CTree test PR and I'd save you a merge conflict

jbduncan replied:
Oh, cool, I hadn't considered that, thank you very much! Let's deal with this then after this PR and the CTree one are merged. :)

tomnelson commented:
I reverted the code using 'truth'. Let's look at the event model for now and later on, decide if it is worth the trouble to create the custom Subjects needed to compare things like Maps of Points and java.awt.Shapes

jbduncan commented:
Sorry @tomnelson, I'm a bit lost. Can you explain to me why you reverted the Truth related changes?

tomnelson commented:
It was not doing what I initially thought, and I removed the commit so as not to clutter up the main work in the PR.
One of my tests was to compare 2 maps of Node to Point, where Point is a custom object with 2 doubles. The API for the truth does not give me a method to do that without creating custom Subjects. The other case was the comparison of two Rectangles, a custom object with 4 doubles. Even comparing 2 arrays of the doubles with equals was not correct. In RTreeTests, when comparing the 2 arrays of doubles, this does not compile:
`      assertWithMessage(
              "The compared bounds are not close enough: rootBounds: %s, unionBounds: %s",
              rootBounds, union)
          .that(rootDimensions)
          .usingTolerance(1.0E-4)
          .isEqualTo(unionDimensions);   
`
Let me know if you find a nice work-around.

jbduncan commented:
Hmm... with regards to comparing maps of Node to Point, if we assume that the classes look like this:

```java
class Node {
  Node(String value) { this.value = value; }
  String value;
  // assuming equals() and hashCode() properly implemented
}

class Point {
  Point(int x, int y) { this.x = x; this.y = y; }
  int x;
  int y;
  // assuming equals() and hashCode() properly implemented
}
```

have you tried using Truth like this?

```java
Map<Node, Point> actual =
    Stream.of(new Node("Hello")).collect(toMap(n -> n, n -> new Point(0, n.value.length())));
Map<Node, Point> expected = ImmutableMap.of(new Node("Hello"), new Point(0, 5));
assertThat(actual).isEqualTo(expected);
```

jbduncan commented:
And for the 2-arrays-of-doubles case, have you tried something like this?

```java
double[] actual = ...;
double[] expected = ...;
assertThat(actual)
        .usingTolerance(1.0E-4)
        .containsExactly(expected)
        .inOrder();
```

jbduncan commented:
> It was not doing what I initially thought, and I removed the commit so as not to clutter up the main work in the PR.

Okay, I'm reasonably happy for us to translate the assertions to Truth at a later time. I opened https://github.com/jrtom/jung/issues/188 yesterday to track this task.

tomnelson commented:
Nice!
That works, as does:

      assertWithMessage(
              "The compared bounds are not close enough: rootBounds: %s,
unionBounds: %s",
              rootBounds, union)
          .that(rootDimensions)
          .usingTolerance(1.0E-4)
          .containsExactly(unionDimensions)
              .inOrder();


On Sun, Jan 14, 2018 at 5:55 PM, Jonathan Bluett-Duncan <
notifications@github.com> wrote:

> And for the 2-arrays-of-doubles case, have you tried something like this?
>
> double[] actual = ...;double[] expected = ...;
> assertThat(actual)
>         .usingTolerance(1.0E-4)
>         .containsExactly(expected)
>         .inOrder();
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrtom/jung/pull/187#issuecomment-357549854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGbB3mbAwEd9HuMNfSSEHoWbxkr72sFks5tKoXsgaJpZM4Rdd15>
> .
>

tomnelson commented:
well, the original intent was to compare the mapped point values within
some tolerance.

On Sun, Jan 14, 2018 at 5:51 PM, Jonathan Bluett-Duncan <
notifications@github.com> wrote:

> Hmm... with regards to comparing maps of Node to Point, if we assume that
> the classes look like this:
>
> class Node {
>   Node(String value) { this.value = value; }
>   String value;
>   // assuming equals() and hashCode() properly implemented
> }
> class Point {
>   Point(int x, int y) { this.x = x; this.y = y; }
>   int x;
>   int y;
>   // assuming equals() and hashCode() properly implemented
> }
>
> have you tried using Truth like this?
>
> Map<Node, Point> actual =
>     Stream.of(new Node("Hello")).collect(toMap(n -> n, n -> new Point(0, n.value.length())));Map<Node, Point> expected = ImmutableMap.of(new Node("Hello"), new Point(0, 5));
> assertThat(actual).isEqualTo(expected);
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrtom/jung/pull/187#issuecomment-357549574>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGbB5_ibPzW0SXZTw6MoDTzt5_e2F7Jks5tKoTwgaJpZM4Rdd15>
> .
>

jbduncan commented:
> well, the original intent was to compare the mapped point values within
some tolerance.

What do you mean by this, exactly?

tomnelson commented:
instead of p.x == q.x i want to ensure that p.x is within some tolerance of
q.x

Math.abs(p.x - q.x) < 0.01  (for example)

(same with the y values)


On Sun, Jan 14, 2018 at 6:03 PM, Jonathan Bluett-Duncan <
notifications@github.com> wrote:

> well, the original intent was to compare the mapped point values within
> some tolerance.
>
> What do you mean by this, exactly?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrtom/jung/pull/187#issuecomment-357550352>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGbB7r3UvWx9DTqI_LQP59pOSR4ISa1ks5tKoe5gaJpZM4Rdd15>
> .
>

tomnelson commented:
Also, if I have to write my own implementations for the test code, i feel
like my test starts to be more about testing my implementations (of things
like PointMapSubject)

On Sun, Jan 14, 2018 at 6:35 PM, tomnelson <notifications@github.com> wrote:

> instead of p.x == q.x i want to ensure that p.x is within some tolerance of
> q.x
>
> Math.abs(p.x - q.x) < 0.01 (for example)
>
> (same with the y values)
>
>
> On Sun, Jan 14, 2018 at 6:03 PM, Jonathan Bluett-Duncan <
> notifications@github.com> wrote:
>
> > well, the original intent was to compare the mapped point values within
> > some tolerance.
> >
> > What do you mean by this, exactly?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/jrtom/jung/pull/187#issuecomment-357550352>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AAGbB7r3UvWx9DTqI_
> LQP59pOSR4ISa1ks5tKoe5gaJpZM4Rdd15>
> > .
>
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jrtom/jung/pull/187#issuecomment-357552494>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGbByh7hRTiQLsVoRWhuDXDGLM5H6H3ks5tKo8xgaJpZM4Rdd15>
> .
>

tomnelson commented:
I fixed a bug when mouse picking was re-activating spatials that should stay inactive during relax phase. Added a demo (based on a redesigned ShowLayouts) where, when a BH using LayoutAlgorithm is selected, the tree structure is drawn under the graph.

tomnelson commented on jung-algorithms/src/main/java/edu/uci/ics/jung/layout/spatial/Node.java:
```diff
@@ -84,6 +84,11 @@ public void insert(ForceObject<T> element) {
+      this.area = area.add(element.p.x, element.p.y);
+    }
+
     if (forceObject == null) {
```
if a node is dragged beyond the original bounds, need to update the bounds to include that location

tomnelson commented on jung-algorithms/src/main/java/edu/uci/ics/jung/layout/spatial/Rectangle.java:
```diff
@@ -52,4 +52,30 @@ public boolean contains(Point p) {
+    double y1 = Math.min(y, newY);
+    double y2 = Math.max(maxY, newY);
+    return new Rectangle(x1, y1, x2 - x1, y2 - y1);
+  }
```
added support to union the rectangle with a new point

tomnelson commented on jung-samples/src/main/java/edu/uci/ics/jung/samples/NodeCollapseDemo.java:
```diff
@@ -159,6 +160,8 @@ public NodeCollapseDemo() {
             vv.getPickedNodeState().clear();
+            //            layoutModel.getLayoutStateChangeSupport().fireLayoutStateChanged(layoutModel, false);
+
             vv.repaint();
```
yeah, i see it. will remove later

tomnelson commented on jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java:
```diff
@@ -442,15 +429,14 @@ protected void renderGraph(Graphics2D g2d) {
-  }
+  //  @Override
+  //  public void layoutStateChanged(LayoutStateChange.Event evt) {}
 
```
mode cruft to remove

jbduncan commented:
@tomnelson 

> Also, if I have to write my own implementations for the test code, i feel
like my test starts to be more about testing my implementations (of things
like PointMapSubject)

Yeah, it's not entirely clear to me either that writing implementations of `Subject` will help us yet, so I think using bog-standard Truth assertions is good enough for now.

tomnelson commented:
@jbduncan you may be interested (amused?) that I created a PointSubject and a Rectangle2DSubject that seem to work pretty well. I would need more unit tests just for the *Subject extensions. Perhaps I will put up a PR with them after your PR using Truth is merged.

jbduncan commented:
> @jbduncan you may be interested (amused?) that I created a PointSubject and a Rectangle2DSubject that seem to work pretty well.

I am indeed interested! (And I admit that I'm also on the verge of chuckling. :wink:)

I'd be interested to at least have a look at these `Subject`s once my CTreeTests PR is merged. I might not be able to do a thorough review, as I'm currently going through job interviews, but I should at least be able to give some hopefully constructive feedback if any is needed. :)

tomnelson commented on jung-visualization/src/test/java/edu/uci/ics/jung/visualization/spatial/rtree/RTreeTest.java:
```diff
@@ -227,25 +229,15 @@ private void testAreas(TreeNode rootNode) {
+      assertThat(rootBounds.getMinY()).isWithin(1.0E-3).of(unionBounds.getMinY());
+      assertThat(rootBounds.getMaxX()).isWithin(1.0E-3).of(unionBounds.getMaxX());
+      assertThat(rootBounds.getMaxY()).isWithin(1.0E-3).of(unionBounds.getMaxY());
     }
```
using minX, maxX, minY, maxY instead of width and height, because Rectangle2D can have negative width/height. I care only that the overall dimensions are close enough that there is no gross error.

tomnelson commented:
closing and squashing commits.
